cdap-data-fabric/src/main/java/co/cask/cdap/store/DefaultNamespaceStore.java

```diff
@@ -48,123 +47,115 @@
  */
 public class DefaultNamespaceStore implements NamespaceStore {
 
-  private static final Id.DatasetInstance APP_META_INSTANCE_ID =
-    Id.DatasetInstance.from(Id.Namespace.SYSTEM, Constants.AppMetaStore.TABLE);
+  private static final DatasetId APP_META_INSTANCE_ID = NamespaceId.SYSTEM.dataset(Constants.AppMetaStore.TABLE);
 
-  private final Supplier<NamespaceMDS> apps;
-  private final Supplier<TransactionExecutor> appsTx;
   private final DatasetFramework dsFramework;
-  private final MultiThreadDatasetCache dsCache;
+  private final Transactional transactional;
 
   @Inject
-  DefaultNamespaceStore(
-    final TransactionExecutorFactory txExecutorFactory,
-    TransactionSystemClient txClient,
-    DatasetFramework framework) {
+  DefaultNamespaceStore(TransactionSystemClient txClient, DatasetFramework framework) {
     this.dsFramework = framework;
-    this.dsCache = new MultiThreadDatasetCache(
-      new SystemDatasetInstantiator(dsFramework), txClient, NamespaceId.SYSTEM, null, null, null);
-    this.apps =
-      new Supplier<NamespaceMDS>() {
-        @Override
-        public NamespaceMDS get() {
-          Table table;
-          try {
-            table = dsCache.getDataset(APP_META_INSTANCE_ID.getId());
-          } catch (DatasetInstantiationException e) {
-            try {
-              DatasetsUtil.getOrCreateDataset(
-                dsFramework, APP_META_INSTANCE_ID, "table",
-                DatasetProperties.EMPTY, DatasetDefinition.NO_ARGUMENTS, null);
-              table = dsCache.getDataset(APP_META_INSTANCE_ID.getId());
-            } catch (DatasetManagementException | IOException e1) {
-              throw Throwables.propagate(e1);
-            }
-          }
-          return new NamespaceMDS(table);
-        }
-      };
-    this.appsTx = new Supplier<TransactionExecutor>() {
-      @Override
-      public TransactionExecutor get() {
-        return txExecutorFactory.createExecutor(ImmutableList.of((TransactionAware) apps.get()));
-      }
-    };
+    this.transactional = Transactions.createTransactional(
+      new MultiThreadDatasetCache(new SystemDatasetInstantiator(dsFramework), txClient,
+                                  NamespaceId.SYSTEM, null, null, null)
+    );
+  }
+
+  private NamespaceMDS getNamespaceMDS(DatasetContext datasetContext) throws IOException, DatasetManagementException {
+    Table table = DatasetsUtil.getOrCreateDataset(datasetContext, dsFramework, APP_META_INSTANCE_ID,
+                                                  Table.class.getName(), DatasetProperties.EMPTY);
+    return new NamespaceMDS(table);
```

Comment:
It is ugly and it was the convention adapted by most Stores. Changing it requires upgrade since type would change from `Table` to `NamespaceMDS`. It's not part of the intend of this PR.